### PR TITLE
Adapt for Android O

### PR DESCRIPTION
**Pull request by MatkovIvan**
- Fixed install update feature on Android O;
- Remove deprecated `singleLine` attribute usage;
- Updated dependencies;
- Fixed infinite recursion in `HockeyLog`;
- Fixed password field font style;
- Fixed/suppressed lint warnings;

### CODE REVIEW

File: hockeysdk/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -6,6 +6,7 @@
     <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
     <uses-permission android:name="android.permission.INTERNET" />
     <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
+    <uses-permission android:name="android.permission.REQUEST_INSTALL_PACKAGES" />
```
Comment (TroubleMakerBen) [inline hockeysdk/src/main/AndroidManifest.xml]: Just an FYI for @ranterle – we're bumping minAPI to 15, and in that, trying to get rid of WRITE_EXTERNAL_STORAGE.  The new permission is required on apiLevel 25+

File: hockeysdk/src/main/java/net/hockeyapp/android/tasks/AttachmentDownloader.java [hunk 2]
```diff
@@ -126,7 +126,7 @@ public boolean hasRetry() {
         }
 
         public boolean consumeRetry() {
-            return --remainingRetries < 0 ? false : true;
+            return --remainingRetries >= 0;
```
Comment (TroubleMakerBen) [inline hockeysdk/src/main/java/net/hockeyapp/android/tasks/AttachmentDownloader.java]: the previous code had pretty bad readability. The new one is better, but not much. Maybe we can be a little more expressive than `--remainingRetries >=0`?

**Comment by MatkovIvan** [comment]
- Bumped min sdk to API level 15;
- Removed fallbacks for unsupported versions;
- `WRITE_EXTERNAL_STORAGE` removed since API 19;
- Changed directories for screenshots and apk download for avoid ask write permission;
- Clean up code from warnings;
- Reused `UpdateFragment` in `UpdateActivity`;
- Open unknown apps settings when install from unknown sources is disabled on Android O;

PS Build fails since missing translations

**Comment by TroubleMakerBen** [comment]
Looks good, I need to test some more on a real device before merging.

File: build.gradle [hunk 3]
```diff
@@ -13,6 +13,9 @@ buildscript {
 allprojects {
     repositories {
         jcenter()
+        maven {
+            url "https://maven.google.com"
```
Comment (guperrot) [inline build.gradle]: required only for early access to v26 and future versions when they are in beta, but 25.3.1 does not require this

File: hockeysdk/src/main/java/net/hockeyapp/android/ExceptionHandler.java [hunk 4]
```diff
@@ -218,7 +217,7 @@ private static void writeValueToFile(String value, String filename) throws IOExc
                 writer.flush();
             }
         } catch (IOException e) {
-            // TODO: Handle exception here
+            e.printStackTrace();
```
Comment (guperrot) [inline hockeysdk/src/main/java/net/hockeyapp/android/ExceptionHandler.java]: Why not using HockeyLog?

File: build.gradle (hunk 3 above)
Comment (TroubleMakerBen) [inline build.gradle]: Good catch!

File: hockeysdk/src/main/java/net/hockeyapp/android/ExceptionHandler.java (hunk 4 above)
Comment (MatkovIvan) [inline hockeysdk/src/main/java/net/hockeyapp/android/ExceptionHandler.java]: [Cause it used widely here.](https://github.com/bitstadium/HockeySDK-Android/search?utf8=%E2%9C%93&q=printStackTrace&type=) Should I change all entries?

Comment (TroubleMakerBen) [inline hockeysdk/src/main/java/net/hockeyapp/android/ExceptionHandler.java]: @MatkovIvan yes, please. And use the error level for it.

**Comment by friederbluemle** [comment]
This PR contains a lot of cleanups and small fixes, which are great, but not directly related to Android O adaption. It would be nice to have them as a separate PR, so it can be merged first, significantly bringing down the lines changed in this PR. This in turn would help with the code review by being able to focus on the things related to Android O.

**Comment by MatkovIvan** [comment]
@friederbluemle Good point, but lot of cleanup caused by raise min API to 15.
@TroubleMakerBen Should we have a separate PR to go some cleanups first?

**Comment by TroubleMakerBen** [comment]
Good point @friederbluemle =)

@MatkovIvan if you can separate cleanup and Android O work within a reasonable timeframe (hals a day max), that'd be great, if it's too much of a hassle, let's keep it as is.

**Comment by friederbluemle** [comment]
Since @MatkovIvan did a great job breaking up the changes into small, logical commits with useful commit messages, it should only take a few minutes to move most of them over to a new branch. Almost all should apply cleanly on the target branch. E.g.
```
git checkout -b fixes-and-cleanups develop   # create new branch off of develop
git cherry-pick 8e32665                      # Fix infinite recursion in HockeyLog
git cherry-pick 55aaf91                      # Update powermock
// etc.
```
Then push the new branch and open another PR. Once merged, rebase `feature/android-o`, do a force-push.

**Comment by friederbluemle** [comment]
@MatkovIvan Please **rebase** your branch (to remove the merge commit and all the other commits that have *already* been merged to `master` via #266).

**Comment by MatkovIvan** [comment]
@friederbluemle It's great that you care about git history, but unfortunately it produces lot of conflicts for most of commits and required double work that already have done in merge commits. Also some commits not really identical since it have been changed in #266 to apply to source without changes related to Android O. So, I don't think it'd be worth the time.

**Comment by TroubleMakerBen** [comment]
FYI I am not a big fan of *rebase* as you can see through all HockeyApp SDKs ;). We have different preferences within the team, though. Anyhow, thanks for participating in all of this! =)

**Comment by friederbluemle** [comment]
You both seem to have pretty big concerns regarding Git rebase. :) I definitely understand different teams have different preferences. When used correctly, rebase (including fixup and squash commits) can have massive benefits for project maintenance. And yes, while it does provide a cleaner history (preventing the slight "spaghettification" seen below), it is certainly not the main (or only) reason. Amongst other things, it can be beneficial for more advanced Git commands (e.g. bisect), as well as simplifying code reviews.

```
$ git log --graph --oneline --no-decorate -24
*   d62430f Merge pull request #268 from MatkovIvan/feature/async-io
|\
| * 5fc23c0 Fix spelling
| * cea471f Fix connection check
| *   9914e74 Merge branch 'feature/android-o' into feature/async-io
| |\
| * \   9a1a130 Merge branch 'fix/cleanup' into feature/async-io
| |\ \
| * | | 3a62598 Remove support-annotations dependency
| * | | e842c83 Make io operations async in FeedbackManager.takeScreenshot
| * | | 38a4274 Make io operations async in PaintActivity
| * | | 97b7f4f Make io operations async in AttachmentView
| * | | 59c3746 Make all io operations async in FeedbackActivity
| * | | c4312f2 Make io operations async in LoginManager.verifyLogin
| * | |   8b3ed6d Merge branch 'fix/cleanup' into feature/async-io
| |\ \ \
| * | | | 804e4d7 Fix tests
| * | | | 0784357 Make all io operations async in MetricsManager
| * | | | a995554 Update tests for CrashManager
| * | | | c0f4f23 Make all io operations async in CrashManager
* | | | |   21a536c Merge pull request #260 from MatkovIvan/feature/android-o
|\ \ \ \ \
| | |_|_|/
| |/| | |
| * | | | 4571c42 Fix bad merge
| * | | |   b224d64 Merge branch 'develop' into feature/android-o
| |\ \ \ \
| |/ / / /
|/| / / /
| |/ / /
* | | |   e7e3238 Merge pull request #266 from MatkovIvan/fix/cleanup
|\ \ \ \
| | |_|/
| |/| |
| * | | eed02a3 Update android target for tests on Travis CI
| * | | 2e18962 Remove support-annotations dependency
| | |/
| |/|
| * | 08766ff Fix unstable test
| * | 8769634 Remove deprecated usages in tests
```

**Comment by TroubleMakerBen** [comment]
Thx for the explanation.

File: hockeysdk/src/main/AndroidManifest.xml [hunk 5]
```diff
@@ -5,7 +5,8 @@
 
     <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
     <uses-permission android:name="android.permission.INTERNET" />
-    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
+    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" android:maxSdkVersion="18" />
```
Comment (laalto) [inline hockeysdk/src/main/AndroidManifest.xml]: This change here caused a bug in my app - I have WRITE_EXTERNAL_STORAGE in a manifest file in a dependency module included before Hockey SDK, and in the merged manifest ended up this version with the `maxSdkVersion` filter, making the permission not to be granted nor requestable.

Guessing this can be an issue for other people so this change might be relevant to mention in release notes.

My solution was to use `tools:node="replace"` manifest merging strategy in a manifest later in the dependency tree:

     <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" tools:node="replace"/>

Comment (TroubleMakerBen) [inline hockeysdk/src/main/AndroidManifest.xml]: Hey, @laalto!
Thx for getting in touch and alerting us to this! I'm adding this to the release notes and to our documentation.

Benny

Comment (koral--) [reply hockeysdk/src/main/AndroidManifest.xml]: IMHO it is less invasive to use `tools:remove="android:maxSdkVersion"` instead of `tools:node="replace"` in such case.
Replace will cause that other attributes from low priority manifests will be replaced instead of being merged.
TL;DR
It seems that currently there only 2 attributes other than `maxSdkVersion` but it may change in the future.

Comment (TroubleMakerBen) [reply hockeysdk/src/main/AndroidManifest.xml]: @koral-- awesome suggestion! I'll update our knowledgebase!
